test/object-store/migrations.cpp

```diff
@@ -2651,3 +2651,120 @@ TEST_CASE("migration: Manual") {
                           Catch::Matchers::ContainsSubstring("Property 'object.value' has been removed."));
     }
 }
+
+TEST_CASE("migration nested collection automatic") {
+    TestFile config;
+    config.schema_mode = SchemaMode::Automatic;
+    auto realm = Realm::get_shared_realm(config);
+
+    Schema schema = {
+        {"nested_object", {{"nested_list", {PropertyType::Array, PropertyType::Array | PropertyType::Int}}}}};
+    REQUIRE_UPDATE_SUCCEEDS(*realm, schema, 0);
+
+    SECTION("Remove nested list") {
+        Schema new_schema = remove_property(schema, "nested_object", "nested_list");
+        REQUIRE_UPDATE_SUCCEEDS(*realm, schema, 1);
+    }
+
+    SECTION("Add nested collection") {
+        Schema new_schema =
+            add_property(schema, "nested_object",
+                         {"another_nested_list", {PropertyType::Array, PropertyType::Array | PropertyType::Int}});
+        REQUIRE_UPDATE_SUCCEEDS(*realm, new_schema, 1);
+    }
+
+    SECTION("Reorder property") {
+        Schema new_schema =
+            add_property(schema, "nested_object",
+                         {"another_nested_list", {PropertyType::Array, PropertyType::Array | PropertyType::Int}});
+        REQUIRE_UPDATE_SUCCEEDS(*realm, new_schema, 1);
+        auto& properties = new_schema.find("nested_object")->persisted_properties;
+        std::swap(properties[0], properties[1]);
+        REQUIRE_UPDATE_SUCCEEDS(*realm, new_schema, 2);
+    }
+
+    SECTION("Change property type should fail with schema in automatic mode") {
+        Schema new_schema = set_type(schema, "nested_object", "nested_list", PropertyType::Int);
+        REQUIRE_UPDATE_FAILS(*realm, new_schema,
+                             "Property 'nested_object.nested_list' has been changed from 'array<int>' to 'int'.");
+    }
+
+    SECTION("add nested collection to same object") {
+        Schema new_schema =
+            add_property(schema, "nested_object",
+                         {"another_nested_list", {PropertyType::Array, PropertyType::Array | PropertyType::Int}});
+        REQUIRE_UPDATE_SUCCEEDS(*realm, new_schema, 1);
+    }
+
+    SECTION("add nested collection with nullable type") {
+        Schema schema = {
+            {"nested_object", {{"nested_list", {PropertyType::Array, PropertyType::Array | PropertyType::Int}}}},
+            {"nested_object2",
+             {{"nested_list_optional",
+               {PropertyType::Array, PropertyType::Array | PropertyType::Int | PropertyType::Nullable}}}},
+        };
+        REQUIRE_UPDATE_SUCCEEDS(*realm, schema, 1);
+    }
+
+    SECTION("change inner type") {
+        Schema schema = {
+            {"nested_object",
+             {{"nested_list",
+               {PropertyType::Array, PropertyType::Array | PropertyType::Mixed | PropertyType::Nullable}}}},
+        };
+        REQUIRE_UPDATE_SUCCEEDS(*realm, schema, 1);
+    }
+
+    SECTION("change type of nested collection") {
```

Comment:
This is something I don't fully grasp, it seems in automatic mode we allow swapping the type in this way, probably I am missing some check.